File: bot/src/application/contracts/commands/negotiate_contract.py
"""Negotiate contract command"""
from dataclasses import dataclass
from datetime import datetime
import requests

from pymediatr import Request, RequestHandler

from domain.shared.contract import Contract, ContractTerms, Delivery, Payment
from domain.shared.value_objects import Waypoint
from domain.shared.exceptions import ContractNegotiationError, RateLimitError
from ports.outbound.repositories import IContractRepository
from ports.outbound.api_client import ISpaceTradersAPI


@dataclass(frozen=True)
class NegotiateContractCommand(Request[Contract]):
    """Command to negotiate a new contract"""
    ship_symbol: str
    player_id: int


class NegotiateContractHandler(RequestHandler[NegotiateContractCommand, Contract]):
    """Handler for NegotiateContractCommand"""

    def __init__(
        self,
        contract_repository: IContractRepository,
        api_client_factory
    ):
        self._contract_repo = contract_repository
        self._api_client_factory = api_client_factory

    async def handle(self, request: NegotiateContractCommand) -> Contract:
        """
        Handle negotiate contract command

        Args:
            request: Command with ship symbol and player ID

        Returns:
            New contract entity

        Raises:
            ContractNegotiationError: If negotiation fails
            RateLimitError: If rate limit is exceeded
        """
        # Get API client for player
        api_client = self._api_client_factory(request.player_id)

        # Call API to negotiate contract with error handling
        try:
            response = api_client.negotiate_contract(request.ship_symbol)
        except requests.exceptions.HTTPError as e:
            # Extract error details from response
            if hasattr(e, 'response') and e.response is not None:
                status_code = e.response.status_code

                # Handle rate limit errors specifically
                if status_code == 429:
                    try:
                        error_data = e.response.json()
                        message = error_data.get('error', {}).get('message', 'Rate limit exceeded')
                        raise RateLimitError(f"Rate limit exceeded (429): {message}")
                    except (ValueError, KeyError):
                        raise RateLimitError("Rate limit exceeded (429)")

                # Handle other HTTP errors
                try:
                    error_data = e.response.json()
                    error_code = error_data.get('error', {}).get('code')
                    error_message = error_data.get('error', {}).get('message', 'Unknown error')

                    # Error 4511 (agent already has active contract) - re-raise HTTPError
                    # to allow batch workflow to handle it by fetching existing contract
                    if error_code == 4511:
                        raise

                    raise ContractNegotiationError(
                        f"Contract negotiation failed (HTTP {status_code}, Error {error_code}): {error_message}"
                    )
                except (ValueError, KeyError):
                    # Response not JSON or missing expected fields
                    raise ContractNegotiationError(
                        f"Contract negotiation failed (HTTP {status_code}): {e.response.text}"
                    )
            else:
                # No response attached to exception
                raise ContractNegotiationError(f"Contract negotiation failed: {str(e)}")

        # Extract contract data from response
        contract_data = response['data']['contract']

        # Parse deliveries
        deliveries = []
        for delivery_data in contract_data['terms']['deliver']:
            delivery = Delivery(
                trade_symbol=delivery_data['tradeSymbol'],
                destination=Waypoint(
                    symbol=delivery_data['destinationSymbol'],
                    x=0.0,  # API doesn't provide coordinates in negotiate response
                    y=0.0
                ),
                units_required=delivery_data['unitsRequired'],
                units_fulfilled=delivery_data.get('unitsFulfilled', 0)
            )
            deliveries.append(delivery)

        # Create payment
        payment = Payment(
            on_accepted=contract_data['terms']['payment']['onAccepted'],
            on_fulfilled=contract_data['terms']['payment']['onFulfilled']
        )

        # Create terms
        terms = ContractTerms(
            deadline=datetime.fromisoformat(contract_data['terms']['deadline']),
            payment=payment,
            deliveries=deliveries
        )

        # Create contract
        contract = Contract(
            contract_id=contract_data['id'],
            faction_symbol=contract_data['factionSymbol'],
            type=contract_data['type'],
            terms=terms,
            accepted=contract_data.get('accepted', False),
            fulfilled=contract_data.get('fulfilled', False),
            deadline_to_accept=datetime.fromisoformat(contract_data['deadlineToAccept'])
        )

        # Save new contract
        self._contract_repo.save(contract, request.player_id)

        return contract

File: gobot/internal/application/contract/negotiate_contract.go
package contract

import (
	"context"
	"fmt"

	"github.com/andrescamacho/spacetraders-go/internal/application/common"
	"github.com/andrescamacho/spacetraders-go/internal/domain/contract"
	"github.com/andrescamacho/spacetraders-go/internal/domain/navigation"
	"github.com/andrescamacho/spacetraders-go/internal/domain/player"
	infraPorts "github.com/andrescamacho/spacetraders-go/internal/infrastructure/ports"
)

// NegotiateContractCommand - Command to negotiate a new contract
type NegotiateContractCommand struct {
	ShipSymbol string
	PlayerID   int
}

// NegotiateContractResponse - Response from negotiate contract command
type NegotiateContractResponse struct {
	Contract      *contract.Contract
	WasNegotiated bool // false if existing contract returned (error 4511)
}

// NegotiateContractHandler - Handles negotiate contract commands
type NegotiateContractHandler struct {
	contractRepo contract.ContractRepository
	shipRepo     navigation.ShipRepository
	playerRepo   player.PlayerRepository
	apiClient    infraPorts.APIClient
}

// NewNegotiateContractHandler creates a new negotiate contract handler
func NewNegotiateContractHandler(
	contractRepo contract.ContractRepository,
	shipRepo navigation.ShipRepository,
	playerRepo player.PlayerRepository,
	apiClient infraPorts.APIClient,
) *NegotiateContractHandler {
	return &NegotiateContractHandler{
		contractRepo: contractRepo,
		shipRepo:     shipRepo,
		playerRepo:   playerRepo,
		apiClient:    apiClient,
	}
}

// Handle executes the negotiate contract command
func (h *NegotiateContractHandler) Handle(ctx context.Context, request common.Request) (common.Response, error) {
	cmd, ok := request.(*NegotiateContractCommand)
	if !ok {
		return nil, fmt.Errorf("invalid request type")
	}

	// 1. Get player to retrieve token
	player, err := h.playerRepo.FindByID(ctx, cmd.PlayerID)
	if err != nil {
		return nil, fmt.Errorf("player not found: %w", err)
	}

	// 2. Load ship from repository
	ship, err := h.shipRepo.FindBySymbol(ctx, cmd.ShipSymbol, cmd.PlayerID)
	if err != nil {
		return nil, fmt.Errorf("ship not found: %w", err)
	}

	// 3. Ensure ship is docked (idempotent)
	stateChanged, err := ship.EnsureDocked()
	if err != nil {
		return nil, err
	}

	// 4. If state was changed, call repository to dock via API
	if stateChanged {
		if err := h.shipRepo.Dock(ctx, ship, cmd.PlayerID); err != nil {
			return nil, fmt.Errorf("failed to dock ship: %w", err)
		}
	}

	// 5. Call API to negotiate contract
	result, err := h.apiClient.NegotiateContract(ctx, cmd.ShipSymbol, player.Token)

	// 6. Handle error 4511 - agent already has active contract
	// Note: API client now parses JSON before checking status, so result.ErrorCode is populated even on errors
	if result != nil && result.ErrorCode == 4511 {
		// Fetch existing contract from API
		existingContractData, err := h.apiClient.GetContract(ctx, result.ExistingContractID, player.Token)
		if err != nil {
			return nil, fmt.Errorf("failed to get existing contract: %w", err)
		}

		// Convert to domain entity
		existingContract := h.convertToDomain(existingContractData, cmd.PlayerID)

		// Save to database
		if err := h.contractRepo.Save(ctx, existingContract); err != nil {
			return nil, fmt.Errorf("failed to save existing contract: %w", err)
		}

		return &NegotiateContractResponse{
			Contract:      existingContract,
			WasNegotiated: false,
		}, nil
	}

	// 7. Convert new contract to domain entity
	newContract := h.convertToDomain(result.Contract, cmd.PlayerID)

	// 8. Save contract to database
	if err := h.contractRepo.Save(ctx, newContract); err != nil {
		return nil, fmt.Errorf("failed to save contract: %w", err)
	}

	return &NegotiateContractResponse{
		Contract:      newContract,
		WasNegotiated: true,
	}, nil
}

// convertToDomain converts API contract data to domain entity
func (h *NegotiateContractHandler) convertToDomain(data *infraPorts.ContractData, playerID int) *contract.Contract {
	// Convert deliveries
	deliveries := make([]contract.Delivery, len(data.Terms.Deliveries))
	for i, d := range data.Terms.Deliveries {
		deliveries[i] = contract.Delivery{
			TradeSymbol:       d.TradeSymbol,
			DestinationSymbol: d.DestinationSymbol,
			UnitsRequired:     d.UnitsRequired,
			UnitsFulfilled:    d.UnitsFulfilled,
		}
	}

	// Build contract terms
	terms := contract.ContractTerms{
		Payment: contract.Payment{
			OnAccepted:  data.Terms.Payment.OnAccepted,
			OnFulfilled: data.Terms.Payment.OnFulfilled,
		},
		Deliveries:       deliveries,
		DeadlineToAccept: data.Terms.DeadlineToAccept,
		Deadline:         data.Terms.Deadline,
	}

	// Create contract
	contractEntity, _ := contract.NewContract(
		data.ID,
		playerID,
		data.FactionSymbol,
		data.Type,
		terms,
	)

	// Restore state from API data
	if data.Accepted {
		contractEntity.Accept()
	}
	if data.Fulfilled {
		contractEntity.Fulfill()
	}

	return contractEntity
}
